### PR TITLE
IObjectCreationExpression API Change

### LINKED ISSUE #18115: IObjectCreationExpression API shape
```
    public interface IObjectCreationExpression : IHasArgumentsExpression
    {
        /// <summary>
        /// Constructor to be invoked on the created instance.
        /// </summary>
        IMethodSymbol Constructor { get; }
        /// <summary>
        /// Explicitly-specified member initializers.
        /// </summary>
        ImmutableArray<ISymbolInitializer> MemberInitializers { get; }
    }
```

The shady part is MemberInitializers, which only expose a part of the IOperation tree within the object member initializer.

Let me give an example:
```
Class F
    Public Field As Integer
End Class

Class C
    Public Sub M1()
        Dim x2 = New F() With {.Field = 2}
    End Sub
End Class
```

IOperation tree for object creation:
```
IOperation tree for "Dim x2 = New F() With {.Field = 2}"

IVariableDeclarationStatement (1 variables) (OperationKind.VariableDeclarationStatement)
  IVariableDeclaration: x2 As F (OperationKind.VariableDeclaration)
    Initializer: IObjectCreationExpression (Constructor: Sub F..ctor()) (OperationKind.ObjectCreationExpression, Type: F)
        Member Initializers: IFieldInitializer (Field: F.Field As System.Int32) (OperationKind.FieldInitializerInCreation)
            ILiteralExpression (Text: 2) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 2)
```

IOperation tree for it's child node, object member initializer:
```
IOperation tree for "With {.Field = 2}"

IOperation (OperationKind.None)
```

IOperation tree for one of the child nodes of the member initializer, NamedFieldInitializerSyntax:
```
IOperation tree for ".Field = 2"

IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Int32)
  Left: IFieldReferenceExpression: F.Field As System.Int32 (OperationKind.FieldReferenceExpression, Type: System.Int32)
      Instance Receiver:  (OperationKind.None)
  Right: ILiteralExpression (Text: 2) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 2)
```

I think we either need to change `MemberInitializers` to return the entire IOperation tree for each initializer (which would be the IAssignmentExpression for the field initializer) OR replace MemberInitializers with `InitializerOpt` of new OperationKind.ObjectMemberInitializer, which exposes all the initializer expressions.

### PR DESCRIPTION
**Pull request by mavasani**
**Customer scenario**

All IOperation based analyzers will generate large number of false positives OR not analyze code within object and collection initializers. Given that such a code pattern is very common, this can be a dogfood blocker for IOperation analyzers (FXCop). 

**Bugs this fixes:**

Fixes #18115

There are still 2 remaining bugs affecting this area:
1. #18781: IOperation API shape for collection initializer expressions
2. #19276: Missing Field/Property reference expression nodes in object creation initializer node

**Workarounds, if any**

The analyzer users will have to either suppress all IOperation analyzers on methods containing object/collection initializers or disable these analyzers.

**Risk**

Low. We are just returning the operation nodes as they are in the bound tree rather than synthesizing new symbol initializer nodes.

**Performance impact**

None

**Is this a regression from a previous update?**

No, this was the original IOperation API implementation

**Root cause analysis:**

The underlying reason is that the current API shape of IObjectCreationExpression only exposes member initializers, not collection initializers.

There are couple of changes in this PR:
1. API change: `ImmutableArray<ISymbolInitializer> MemberInitializers` is changed to `ImmutableArray<IOperation> Initializers` as per design time decision.
2. Implementation changes:
   1. Instead of returning the member initializers as synthesized ISymbolInitializer nodes, we now return member initializers as IAssignmentExpression nodes. This ensures completeness of IOperation tree.
   2. Now we also return the collection initializer expressions within an object creation expression.

**How was the bug found?**

Dogfooding.

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation for review.

**Comment by mavasani** [comment]
Also tagging @AlekseyTs and @cston

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb [hunk 1]
```diff
@@ -238,33 +238,48 @@ Friend Class [Class]
     Public Property C As [Class]
 
     Public Sub M(x As Integer, y As Integer, z As Integer)
-        Dim c = New [Class]() With {'BIND:"New [Class]() With {"
+        Dim c = New [Class]() With {'BIND:"New [Class]() With {"'BIND:"New [Class]() With {'BIND:"New [Class]() With {""
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb]: What are the extra `'BIND ...` comments for?

**Comment by cston** [comment]
LGTM

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb (hunk 1 above)
Comment (mavasani) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb]: Seems like they got left behind accidentally when I copied another existing test and modified it. There are few more tests with extra bind comments, I will clean them separately.

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@AlekseyTs Can you take a look? Thanks.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb [hunk 2]
```diff
@@ -238,33 +238,48 @@ Friend Class [Class]
...
-      IObjectCreationExpression (Constructor: Sub System.Collections.Generic.Dictionary(Of System.Int32, System.Int32)..ctor()) (OperationKind.ObjectCreationExpression, Type: System.Collections.Generic.Dictionary(Of System.Int32, System.Int32)) (Syntax: 'New Diction ... om {{x, y}}')
-    IPropertyInitializer (Property: Property [Class].C As [Class]) (OperationKind.PropertyInitializerInCreation) (Syntax: '.C = New [C ... th {.X = z}')
-      IObjectCreationExpression (Constructor: Sub [Class]..ctor()) (OperationKind.ObjectCreationExpression, Type: [Class]) (Syntax: 'New [Class] ... th {.X = z}')
-        Member Initializers(1): IPropertyInitializer (Property: Property [Class].X As System.Int32) (OperationKind.PropertyInitializerInCreation) (Syntax: '.X = z')
-            IParameterReferenceExpression: z (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'z')
+  Initializers(4): IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Void) (Syntax: '.X = x')
+      Left: IIndexedPropertyReferenceExpression: Property [Class].X As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'X')
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb]: >Left: IIndexedPropertyReferenceExpression: Property [Class].X As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'X') [](start = 6, length = 155)

It doesn't look like the property is an indexed property. The same issue is with other properties.

Comment (mavasani) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb]: Yes, this is a bug that is already fixed in features/ioperation branch. I also updated the test in that branch accordingly: https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb#L94

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        }/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+IObjectCreationExpression (Constructor: Class..ctor()) (OperationKind.ObjectCreationExpression, Type: Class) (Syntax: 'new Class() ... }')
+  Initializers(4): IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Int32) (Syntax: 'X = x')
+      Left: IOperation:  (OperationKind.None) (Syntax: 'X')
+      Right: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
+    IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Collections.Generic.List<System.Int32>) (Syntax: 'Y = { x, y, 3 }')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: >IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Collections.Generic.List<System.Int32>) (Syntax: 'Y = { x, y, 3 }') [](start = 4, length = 139)

This feels surprising. Does the assignment really happen in this scenario?

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Would you like me to file an issue or are you suggesting we not expose the IAssignmentExpression in the operation tree?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public partial class IOperationTests : SemanticModelTestBase
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: >IOperationTests [](start = 25, length = 15)

It looks like there are no tests for an initializer target in the form of an argument list.

initializer_target
    : identifier
    | '[' argument_list ']'
    ;

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Yes, this PR is not meant to add complete set of tests for IObjectCreationExpression. I will add them later as part of unit testing effort https://github.com/dotnet/roslyn/issues/17588

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: I think opening an issue should be fine at this point.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Filed https://github.com/dotnet/roslyn/issues/19375

**Comment by mavasani** [comment]
Tagging @MattGertz for ask mode approval.

**Pull request merged**